### PR TITLE
Add index add api

### PR DESCRIPTION
**Pull request by BBuf**
- [x] cuda kernel
- [x] unittest
- [x] cpu kernel
- [x] docs

图片

图片

图片

### CODE REVIEW

File: oneflow/core/functional/impl/array_functor.cpp [hunk 1]
```diff
@@ -3371,6 +3372,76 @@ class FillTensorFunctor {
...
+                         .Build());
+  }
+  Maybe<Tensor> operator()(const std::shared_ptr<one::Tensor>& input, const int64_t& dim,
+                           const std::shared_ptr<one::Tensor>& index,
+                           const std::shared_ptr<one::Tensor>& source, const Scalar& alpha) const {
+    const float alpha_value = alpha.As<float>();
+    int64_t dim_ = dim;
+    const auto& input_shape = input->shape();
+    const int64_t& num_axes = input_shape->NumAxes();
+    dim_ = JUST(maybe_wrap_dim(dim_, num_axes));
```
Comment (Flowingsun007) [inline oneflow/core/functional/impl/array_functor.cpp:3425]:
```suggestion
    dim_ = JUST(maybe_wrap_dim(dim_, input->ndim()));
```

File: oneflow/core/functional/impl/array_functor.cpp [hunk 2]
```diff
@@ -3371,6 +3372,76 @@ class FillTensorFunctor {
...
```
[the remark this answers is not in the document]
Comment (Flowingsun007) [reply oneflow/core/functional/impl/array_functor.cpp:3392]: L3390-L3391可以删掉

[the remark this answers is not in the document]
Comment (BBuf) [reply oneflow/core/functional/impl/array_functor.cpp:3392]: 好的

File: oneflow/core/functional/impl/array_functor.cpp [hunk 3]
```diff
@@ -3371,6 +3372,76 @@ class FillTensorFunctor {
...
+    CHECK_OR_RETURN(index->dtype()->data_type() != DataType::kInt32
```
[the remark this answers is not in the document]
Comment (BBuf) [reply oneflow/core/functional/impl/array_functor.cpp:3393]: 好的

File: oneflow/user/kernels/index_add_kernel.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+#define REGISTER_INDEX_ADD_CPU_KERNEL(dtype)                          \
+  REGISTER_USER_KERNEL("index_add")                                   \
+      .SetCreateFn<IndexAddCpuKernel<dtype>>()                        \
+      .SetIsMatchedHob((user_op::HobDeviceType() == DeviceType::kCPU) \
+                       && (user_op::HobDataType("output", 0) == GetDataType<dtype>::value));
+
+REGISTER_INDEX_ADD_CPU_KERNEL(float)
+REGISTER_INDEX_ADD_CPU_KERNEL(double)
```
[the remark this answers is not in the document]
Comment (BBuf) [reply oneflow/user/kernels/index_add_kernel.cpp:81]: 可以的

**Comment by marigoold** [comment]
https://github.com/pytorch/pytorch/blob/51e961dd7bb9abaf999e6028208b2778a57c32b2/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L330-L331

torch 对这类 op 都会有一个 `source.dim() == 0 || numIndices == source.size(dim)` 检查，这里需要对齐吗？

File: python/oneflow/test/modules/test_index_add.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    # check inplace
+    torch_x.index_add_(0, torch_index, torch_t)
+    flow_x.index_add_(0, flow_index, flow_t)
+    test_case.assertTrue(
+        np.allclose(torch_y.cpu().numpy(), flow_y.cpu().numpy(), 1e-05, 1e-05)
+    )
+
+
+@flow.unittest.skip_unless_1n1d()
+class TestIndexAdd(flow.unittest.TestCase):
```
Comment (Flowingsun007) [inline python/oneflow/test/modules/test_index_add.py:58]: 要不要加一个autotest的case?

**Comment by BBuf** [comment]
> https://github.com/pytorch/pytorch/blob/51e961dd7bb9abaf999e6028208b2778a57c32b2/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L330-L331
> 
> torch 对这类 op 都会有一个 `source.dim() == 0 || numIndices == source.size(dim)` 检查，这里需要对齐吗？

可以对齐，我加一下

File: python/oneflow/test/modules/test_index_add.py (hunk 5 above)
Comment (BBuf) [reply python/oneflow/test/modules/test_index_add.py:58]: 想过，主要是不好加

Comment (BBuf) [reply python/oneflow/test/modules/test_index_add.py:58]: 可以，我加一下

File: oneflow/user/kernels/index_add_kernel.cu [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    const int64_t dim = ctx->Attr<int64_t>("dim");
+    const float alpha = ctx->Attr<float>("alpha");
+    const ShapeView& input_shape = input->shape_view();
+    const ShapeView& source_shape = source->shape_view();
+    std::vector<int64_t> input_stride(input->stride().begin(), input->stride().end());
+    const int64_t stride = input_stride[dim];
+    const int64_t source_dim = source_shape.At(dim);
+    const int64_t delta = input_shape.At(dim) - source_dim;
+    DataType index_dtype = index->data_type();
+    const int32_t n = input->shape_view().elem_cnt();
```
Comment (marigoold) [inline oneflow/user/kernels/index_add_kernel.cu:64]: 不知道是不是我理解有误，我觉得这里的 n 应该是 source 的 elem_cnt，比如对于这种 case，这个 kernel 还可以得到正确的结果吗？
```
In [8]: x = torch.ones(2, 3)

In [9]: t = torch.ones(6, 3)

In [10]: index = torch.tensor([0, 1, 0, 1, 0, 1])

In [11]: x.index_add_(0, index, t)
Out[11]:
tensor([[4., 4., 4.],
        [4., 4., 4.]])
```

File: oneflow/user/kernels/index_add_kernel.cu [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+namespace {
+template<typename T, typename IndexT>
+__global__ void index_add_cuda_kernel(const int64_t n, const T* input, const IndexT* index,
+                                      const T* source, T* output, const int64_t stride,
+                                      const int64_t source_dim, const int64_t delta,
+                                      const float alpha) {
+  const int64_t stride_source_dim = stride * source_dim;
+  CUDA_1D_KERNEL_LOOP(i, n) {
+    int64_t pre_index = i / stride_source_dim;
+    int64_t dim_index = i % stride_source_dim / stride;
```
Comment (marigoold) [inline oneflow/user/kernels/index_add_kernel.cu:34]: 这个取余是不是可以改成 `i - stride_source_dim * pre_index` 优化一下？

File: oneflow/user/kernels/index_add_kernel.cu (hunk 6 above)
Comment (BBuf) [reply oneflow/user/kernels/index_add_kernel.cu:64]: 手滑了，感谢指出，已修正

File: oneflow/user/kernels/index_add_kernel.cu (hunk 7 above)
Comment (BBuf) [reply oneflow/user/kernels/index_add_kernel.cu:34]: 好的

Comment (BBuf) [reply oneflow/user/kernels/index_add_kernel.cu]: 应该改成：i - i / stride_source_dim * stride_source_dim 才对

**Review by Flowingsun007 — APPROVED** [review]

File: oneflow/user/kernels/index_add_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  // source_dim_idx = index[dim_index] = index[0] = 0
+  // output_index = i + (delta * pre_index + source_dim_idx - dim_index) * stride = 9 + (2 * 0 + 0 -
+  // 0) * 3 = 9 cuda::atomic::Add(output + output_index, static_cast<T>(alpha) * source[i])=>
+  // output[9] += 1.0 * 9 = 10.0
+  const int64_t stride_source_dim = stride * source_dim;
+  CUDA_1D_KERNEL_LOOP(i, n) {
+    int64_t pre_index = i / stride_source_dim;
+    int64_t dim_index = (i - i / stride_source_dim * stride_source_dim) / stride;
+    IndexT source_dim_idx = index[dim_index];
+    int64_t output_index = i + (delta * pre_index + source_dim_idx - dim_index) * stride;
```
Comment (BBuf) [inline oneflow/user/kernels/index_add_kernel.cu:54]: 这里的反索引可以以2维Tensor也就是矩阵为例来进行推导，上面的注释给了一个例子方便理解。

File: python/oneflow/test/modules/test_index_add.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def _test_index_add(test_case, device):
+    torch_x = torch.ones(5, 3).to(device)
+    torch_t = torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=torch.float).to(
+        device
+    )
+    torch_index = torch.tensor([0, 4, 2]).to(device)
+    torch_y = torch.index_add(torch_x, 0, torch_index, torch_t)
+    torch_y_alpha = torch.index_add(torch_x, 0, torch_index, torch_t, alpha=-1)
```
Comment (Flowingsun007) [inline python/oneflow/test/modules/test_index_add.py:33]: index tensor只能允许int32/int64，input tensor可以是float类型吗？如果可以，建议加一个case

File: oneflow/core/functional/impl/array_functor.cpp [hunk 10]
```diff
@@ -3371,6 +3372,78 @@ class FillTensorFunctor {
...
+        << "Input(Index) holds the wrong type, it holds "
+        << DataType_Name(index->dtype()->data_type())
+        << " , but "
+           "desires to be int32_t or int64_t";
+    const float alpha_value = alpha.As<float>();
+    int64_t dim_ = dim;
+    dim_ = JUST(maybe_wrap_dim(dim_, input->ndim()));
+    auto& attrs = THREAD_CACHED_MUTABLE_ATTR_MAP("dim", "alpha");
+    attrs.SetAllAttrs(dim_, alpha_value);
+    return OpInterpUtil::Dispatch<Tensor>(*op_, {input, index, source}, attrs);
```
Comment (wyg1997) [inline oneflow/core/functional/impl/array_functor.cpp:3402]: 这里 input 和 index 可能需要用 TensorProcessor 处理一下，加一个 common_dtype 的 flag

File: python/oneflow/test/modules/test_index_add.py (hunk 9 above)
Comment (BBuf) [reply python/oneflow/test/modules/test_index_add.py:33]: 这个case就是float32的。

File: python/oneflow/test/modules/test_index_add.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    torch_x.index_add_(0, torch_index, torch_t)
+    flow_x.index_add_(0, flow_index, flow_t)
+    test_case.assertTrue(
+        np.allclose(torch_y.cpu().numpy(), flow_y.cpu().numpy(), 1e-05, 1e-05)
+    )
+
+
+@flow.unittest.skip_unless_1n1d()
+class TestIndexAdd(flow.unittest.TestCase):
+    def test_index_add(test_case):
```
Comment (daquexian) [inline python/oneflow/test/modules/test_index_add.py:59]: 加一个 profile 测试吧~

File: python/oneflow/test/modules/test_index_add.py (hunk 9 above)
Comment (Flowingsun007) [reply python/oneflow/test/modules/test_index_add.py:33]: ok

Comment (Flowingsun007) [reply python/oneflow/test/modules/test_index_add.py:33]: 那index tensor是不是没有测int32或int64的？

Comment (BBuf) [reply python/oneflow/test/modules/test_index_add.py:33]: 不需要测试，index_add cuda kernel目前有atomic add，是只能支持input为half，fp32，fp64，这个和pytorch是对齐的。

File: python/oneflow/test/modules/test_index_add.py (hunk 11 above)
Comment (BBuf) [reply python/oneflow/test/modules/test_index_add.py:59]: 好的

File: python/oneflow/framework/docstr/index_add.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+                [ 1.,  1.,  1.],
+                [ 8.,  9., 10.],
+                [ 1.,  1.,  1.],
+                [ 5.,  6.,  7.]], dtype=oneflow.float32)
+        >>> x.index_add_(0, index, t, alpha=-1)
+        tensor([[1., 1., 1.],
+                [1., 1., 1.],
+                [1., 1., 1.],
+                [1., 1., 1.],
+                [1., 1., 1.]], dtype=oneflow.float32)
```
Comment (wyg1997) [inline python/oneflow/framework/docstr/index_add.py:68]: 这个例子是不是弄错了，alpha 是 -1 的话，结果不应该是这样的

Comment (BBuf) [reply python/oneflow/framework/docstr/index_add.py:68]: 没有吧。https://pytorch.org/docs/stable/generated/torch.Tensor.index_add_.html#torch.Tensor.index_add_

Comment (BBuf) [reply python/oneflow/framework/docstr/index_add.py:68]: 图片

Comment (BBuf) [reply python/oneflow/framework/docstr/index_add.py:68]: 我看了一下，这个地方是因为原地更新导致x的值修改了一次，然后又修改回去了，最后就变成全1 了。

Comment (wyg1997) [reply python/oneflow/framework/docstr/index_add.py:68]: 哦哦哦，是因为连续调用又减回去了，我看错了

File: oneflow/user/kernels/index_add_kernel.cu [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+#define REGISTER_INDEX_ADD_CUDA_KERNEL(dtype)                          \
+  REGISTER_USER_KERNEL("index_add")                                    \
+      .SetCreateFn<IndexAddGpuKernel<dtype>>()                         \
+      .SetIsMatchedHob((user_op::HobDeviceType() == DeviceType::kCUDA) \
+                       && (user_op::HobDataType("output", 0) == GetDataType<dtype>::value));
+
+REGISTER_INDEX_ADD_CUDA_KERNEL(float)
+REGISTER_INDEX_ADD_CUDA_KERNEL(half)
+REGISTER_INDEX_ADD_CUDA_KERNEL(double)
+
```
Comment (wyg1997) [inline oneflow/user/kernels/index_add_kernel.cu:108]: 类型注册这里，cpu 和 cuda 注册的类型不一样，这里有什么特别的原因吗？

Comment (BBuf) [reply oneflow/user/kernels/index_add_kernel.cu:108]: 这里要用到atomic add所以cuda kernel里面只能支持half,float32/64，cpu里面的+=是可以支持int的，所以我就注册了int。

File: oneflow/user/ops/index_add_op.cpp [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  ctx->SetOutputStride("output", 0, ctx->InputStride("input", 0));
+  return Maybe<void>::Ok();
+}
+
+/*static*/ Maybe<void> IndexAddOp::InferPhysicalTensorDesc(user_op::InferContext* ctx) {
+  return InferLogicalTensorDesc(ctx);
+}
+
+/* static */ Maybe<void> IndexAddOp::GetSbp(user_op::SbpContext* ctx) {
+  // TODO(yangzhimin): support more valid sbp signature.
```
Comment (wyg1997) [inline oneflow/user/ops/index_add_op.cpp:33]: @small1945 这里后面你可以来完善一下 sbp 推导和 global 单测

File: oneflow/user/kernels/index_add_kernel.cu (hunk 13 above)
Comment (wyg1997) [reply oneflow/user/kernels/index_add_kernel.cu:108]: 哦哦好的

File: python/oneflow/test/modules/test_index_add.py (hunk 11 above)
Comment (BBuf) [reply python/oneflow/test/modules/test_index_add.py:65]: 图片

性能差不多，这个kernel还有优化机会，我后面研究一下。

File: oneflow/user/ops/index_add_op.cpp (hunk 14 above)
Comment (small1945) [reply oneflow/user/ops/index_add_op.cpp:33]: 收到

File: python/oneflow/test/modules/test_as_strided.py [hunk 15]
```diff
@@ -44,6 +44,26 @@ def test_flow_AsStrided(test_case):
...
+        dim1 = np.random.randint(2, 4)
+        dim2 = np.random.randint(2, 4)
+        dim3 = np.random.randint(2, 4)
+        dim4 = np.random.randint(2, 4)
+        if ndim == 3:
+            x = random_tensor(3, dim0, dim1, dim2)
+        elif ndim == 4:
+            x = random_tensor(4, dim0, dim1, dim2, dim3)
+        elif ndim == 5:
+            x = random_tensor(5, dim0, dim1, dim2, dim3, dim4)
```
Comment (wyg1997) [inline python/oneflow/test/modules/test_as_strided.py:61]:
```suggestion
        ndim = np.random.randint(3, 6)
        x = random_tensor(ndim, *[np.random.randint(2, 4) for _ in range(ndim)])
```

可以好像可以简化一下

**Review by wyg1997 — APPROVED** [review]

File: python/oneflow/test/modules/test_as_strided.py (hunk 15 above)
Comment (BBuf) [reply python/oneflow/test/modules/test_as_strided.py:61]: 有道理，我改下吧，这是春游的杰作
